flexmeasures/auth/decorators.py

```diff
@@ -168,38 +172,68 @@ def post(self, resource_data: dict):
-            if ctx_arg_pos is not None and ctx_arg_name is not None:
-                context_from_args = args[ctx_arg_pos][ctx_arg_name]
-            elif ctx_arg_pos is not None:
-                context_from_args = args[ctx_arg_pos]
-            elif ctx_arg_name is not None:
-                context_from_args = kwargs[ctx_arg_name]
-            elif len(args) > 0:
-                context_from_args = args[0]
-
-            # if a loader is given, use that, otherwise fall back to context_from_args
-            if ctx_loader is not None:
-                if pass_ctx_to_loader:
-                    if inspect.isclass(ctx_loader) and issubclass(
-                        ctx_loader, AuthModelMixin
-                    ):
-                        context = db.session.get(ctx_loader, context_from_args)
-                    else:
-                        context = ctx_loader(context_from_args)
-                else:
-                    context = ctx_loader()
-            else:
-                context = context_from_args
+            context = load_context(
+                ctx_arg_pos, ctx_arg_name, ctx_loader, pass_ctx_to_loader, args, kwargs
+            )
 
-            check_access(context, permission)
+            # skip check in case (optional) argument was not passed
+            if context is None:
+                return fn(*args, **kwargs)
+
+            # Check access for possibly multiple contexts
+            if not isinstance(context, list):
+                context = [context]
+            for ctx in context:
+                if isinstance(ctx, tuple):
+                    c = ctx[0]  # c[0] is the context, c[1] is its origin
+                    # the context loader may narrow down the origin of the context (e.g. a nested field rather than a function argument)
+                    origin = ctx[1]
+                else:
+                    c = ctx
+                    origin = ctx_arg_name
+                try:
+                    check_access(c, permission)
+                except Exception as e:  # noqa: B902
+                    if error_handler:
+                        return error_handler(c, permission, origin)
+                    raise e
 
             return fn(*args, **kwargs)
 
         return decorated_view
 
     return wrapper
+
+
+def load_context(
+    ctx_arg_pos, ctx_arg_name, ctx_loader, pass_ctx_to_loader, args, kwargs
+) -> AuthModelMixin | None:
+
+    # first set context_from_args, if possible
+    context_from_args: AuthModelMixin | None = None
+    if ctx_arg_pos is not None and ctx_arg_name is not None:
+        context_from_args = args[ctx_arg_pos][ctx_arg_name]
+    elif ctx_arg_pos is not None:
+        context_from_args = args[ctx_arg_pos]
+    elif ctx_arg_name is not None:
+        context_from_args = kwargs.get(ctx_arg_name)
+        # skip check in case (optional) argument was not passed
+        if context_from_args is None:
+            return None
```

Comment:
I think this is the major problem blocking my progress on this PR. How do you propose I deal with checking authorization on an optional argument? Should I add a boolean argument to `permission_required_for_context` telling the argument is optional (i.e. **skip** if context is None) or required (i.e. **raise** if context is None)?

Reply:
Can we not check that `context_from_args` has to be something if the developer gave information to look it up?

I.e. if `ctx_arg_name` is not `None`, then we expect that `context_from_args` is also not `None`, otherwise we got wrong parameterisation.

Simply put, if `permission_required_for_context()` has no context, we cannot say the permission is granted.

Why did you change the logic so that it could be granted? The field was optional before. Maybe it is because the flex context can contain no sensors to check? That is a case I can imagine, and maybe we could add a parameter to `permission_required_for_context()` that says `pass_if_no_context_found` or similar.

But the interpretation has to fit.